File: frontend/src/utils/hostsParser.js
function isValidIPv4(ip) {
  const parts = ip.split(".");
  if (parts.length !== 4) {
    return false;
  }

  return parts.every((part) => {
    if (!/^\d+$/.test(part)) {
      return false;
    }

    const value = Number(part);
    return value >= 0 && value <= 255;
  });
}

function isLikelyIPv6(ip) {
  if (!ip.includes(":")) {
    return false;
  }

  return /^[0-9a-fA-F:.]+$/.test(ip);
}

function isValidIP(ip) {
  return isValidIPv4(ip) || isLikelyIPv6(ip);
}

function parseHostsLine(rawLine, lineNumber) {
  const trimmed = rawLine.trim();
  if (!trimmed) {
    return null;
  }

  let disabled = false;
  let lineToParse = trimmed;

  if (lineToParse.startsWith("#")) {
    disabled = true;
    lineToParse = lineToParse.slice(1).trim();
    if (!lineToParse) {
      return null;
    }
  }

  let comment = "";
  const commentIndex = lineToParse.indexOf("#");
  if (commentIndex >= 0) {
    comment = lineToParse.slice(commentIndex + 1).trim();
    lineToParse = lineToParse.slice(0, commentIndex).trim();
  }

  if (!lineToParse) {
    return null;
  }

  const fields = lineToParse.split(/\s+/).filter(Boolean);
  if (fields.length < 2) {
    if (disabled) {
      return null;
    }

    throw new Error(`Invalid hosts entry on line ${lineNumber}: ${rawLine}`);
  }

  const ip = fields[0];
  if (!isValidIP(ip)) {
    if (disabled) {
      return null;
    }

    throw new Error(`Invalid IP address on line ${lineNumber}: ${ip}`);
  }

  return {
    line: lineNumber,
    raw: rawLine,
    ip,
    hostnames: fields.slice(1),
    comment,
    disabled,
  };
}

export function parseHostsContent(content) {
  const lines = content.split(/\r?\n/);
  const entries = [];

  lines.forEach((line, index) => {
    const entry = parseHostsLine(line, index + 1);
    if (entry) {
      entries.push(entry);
    }
  });

  return {
    entries,
  };
}

File: hosts/hostsParser.go
package hosts

import (
	"bufio"
	"encoding/json"
	"fmt"
	"net"
	"os"
	"strings"
)

// HostsFile is the JSON-friendly representation of a hosts file.
type HostsFile struct {
	Entries []HostEntry `json:"entries"`
}

// HostEntry represents a single hosts mapping line.
type HostEntry struct {
	Line      int      `json:"line"`
	Raw       string   `json:"raw"`
	IP        string   `json:"ip"`
	Hostnames []string `json:"hostnames"`
	Comment   string   `json:"comment,omitempty"`
	Disabled  bool     `json:"disabled"`
}

// ParseHosts converts raw hosts file content into a structured representation.
func ParseHosts(content string) (HostsFile, error) {
	scanner := bufio.NewScanner(strings.NewReader(content))
	result := HostsFile{Entries: []HostEntry{}}

	lineNumber := 0
	for scanner.Scan() {
		lineNumber++

		entry, ok, err := parseHostsLine(scanner.Text(), lineNumber)
		if err != nil {
			return HostsFile{}, err
		}
		if ok {
			result.Entries = append(result.Entries, entry)
		}
	}

	if err := scanner.Err(); err != nil {
		return HostsFile{}, fmt.Errorf("scan hosts content: %w", err)
	}

	return result, nil
}

// ParseHostsToJSON converts raw hosts file content directly into JSON.
func ParseHostsToJSON(content string) (string, error) {
	parsed, err := ParseHosts(content)
	if err != nil {
		return "", err
	}

	data, err := json.MarshalIndent(parsed, "", "  ")
	if err != nil {
		return "", fmt.Errorf("marshal hosts json: %w", err)
	}

	return string(data), nil
}

// ParseHostsFile reads a hosts file from disk and parses it.
func ParseHostsFile(path string) (HostsFile, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return HostsFile{}, fmt.Errorf("read hosts file: %w", err)
	}

	return ParseHosts(string(content))
}

// ParseHostsFileToJSON reads a hosts file from disk and returns its JSON representation.
func ParseHostsFileToJSON(path string) (string, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return "", fmt.Errorf("read hosts file: %w", err)
	}

	return ParseHostsToJSON(string(content))
}

func parseHostsLine(rawLine string, lineNumber int) (HostEntry, bool, error) {
	trimmed := strings.TrimSpace(rawLine)
	if trimmed == "" {
		return HostEntry{}, false, nil
	}

	disabled := false
	lineToParse := trimmed

	if strings.HasPrefix(lineToParse, "#") {
		disabled = true
		lineToParse = strings.TrimSpace(strings.TrimPrefix(lineToParse, "#"))
		if lineToParse == "" {
			return HostEntry{}, false, nil
		}
	}

	comment := ""
	if index := strings.Index(lineToParse, "#"); index >= 0 {
		comment = strings.TrimSpace(lineToParse[index+1:])
		lineToParse = strings.TrimSpace(lineToParse[:index])
	}

	if lineToParse == "" {
		return HostEntry{}, false, nil
	}

	fields := strings.Fields(lineToParse)
	if len(fields) < 2 {
		if disabled {
			return HostEntry{}, false, nil
		}

		return HostEntry{}, false, fmt.Errorf("invalid hosts entry on line %d: %q", lineNumber, rawLine)
	}

	ipAddress := fields[0]
	if net.ParseIP(ipAddress) == nil {
		if disabled {
			return HostEntry{}, false, nil
		}

		return HostEntry{}, false, fmt.Errorf("invalid IP address on line %d: %q", lineNumber, ipAddress)
	}

	return HostEntry{
		Line:      lineNumber,
		Raw:       rawLine,
		IP:        ipAddress,
		Hostnames: fields[1:],
		Comment:   comment,
		Disabled:  disabled,
	}, true, nil
}
